Get iris-grib working in Python 3 !

lbdreyer opened the pull request:
The eccodes 2.10 release includes Python 3 support.

This PR includes Python 3 in the testing matrix and thus highlights outstanding issues with iris-grib/eccodes/Python3

Known issues with Python 3 that work fine in Python 2:
- [x] **1) Can't set the key 'NV' as np.int64**
   **TESTS:** integration.save_rules.test_save_hybrid_coords (all tests in file)
- [x] **2) Problem with lazy loading of _data_ from files**
    **2.i) files with hybrid factory - problem loading orography field during factory creation**
   files: iris_grib/tests/testdata/faked_sample_hh_grib_data.grib2 and faked_sample_hp_grib_data.grib2
   **TESTS:** integration.load_convert.test_load_hybrid_coords (all tests in file) and integration.round_trip.test_hybrid_coords_round_trip (all tests in file)
    **2.ii)) Can load in file as cube, when trying to call the data it fails**
  files: iris_grib/tests/testdata/*.grib2, (excet CB which does work)
   **TESTS:** integration.round_trip/test_WAFC_mapping_round_trip all tests in file except CB.grib2)

rsignell-usgs commented:
This is our best bet to get GRIB-writing going with Python 3, right?

ocefpaf commented:
> This is our best bet to get GRIB-writing going with Python 3, right?

The alternative is the "raw" python-eccodes, wrapped here, but I do not recommend it. `iris-grib` is much better and provide the high level interface we need.

@bjlittle and @lbdreyer are the errors here to hard to fix? I'm not sure what to make of all those 'grib message errors' there :-/

bjlittle commented:
@rsignell-usgs and @ocefpaf Unfortunately, we can't work around (in iris-grib) the issues that we're hitting due to the problems with ecCodes in Python3. As a result, we've not deployed GRIB support for Python3 here at the Met Office. That's a major blocker. We've escalated this at an organisational level with the ECMWF. As a result, they are scheduled to come to us after Easter to discuss this further and update us on their progress with estimates of when ecCodes will be fully functional in Python3.

HTH

ocefpaf commented:
Thanks for the clarifications @bjlittle! Let me know if I can do anything to help.

lbdreyer commented:
I have changed this PR so that it uses eccodes-python for Python 3 (as eccodes-python is Python 3 only) and uses python-eccodes for Python 2.

It looks like eccodes-python has fixed most of our problems with Python 3. :tada: :tada: :tada:

However, you will notice that to get the tests passing I had to include some unpleasant workarounds. I am definitely not suggesting we use these workarounds, I just wanted to show that it would work. I am going to raise the outstanding issues with the developers of eccodes-python, namely that:
* Sometimes grib returns an array when we expect a list (e.g. `gribapi.grib_get_double_array' returns a list)
* It fails if we try to set the`NV` key with a object of type `numpy.int64` and so we have to convert the Python with `int()`

lbdreyer commented:
I have updated this PR with a better workaround for when `gribapi.get_*array` returns a list as opposed to a numpy array.

This PR now contains two commits:
commit 1: introduces testing with Python3/eccodes-python; adds a few fixes necessary for the latest eccodes.
commit 2: includes the *temporary* wrapper that we can use whenever we call `gribapi.grib_get_*array`
I have split it like this as my intention is to remove the contents of commit 2 once eccodes-python has been fixed (I have raised this issue with the developers of eccodes-python and they are looking into it)

I suggest we merge this in so that we can get users testing with iris-grib/Python3 and potentially find issues our own testing of eccodes-python has not.

lbdreyer commented:
@bjlittle are you able to look at this?

bjlittle commented on .travis.yml:
```diff
@@ -2,6 +2,7 @@ language: python
 sudo: false
 env:
   - python=2.7
+  - python=3.6
```
@lbdreyer Awesome!

bjlittle commented on .travis.yml:
```diff
@@ -16,7 +17,24 @@ install:
+  # If running in Python 3, we will install Python3-only eccodes-python via pip.
+  - if [[ "$python" == "3.6" ]]; then
+        export ECCODES_DEPS="eccodes pip";
+    fi
```
@lbdreyer To simply we could make this the `else` of `"2.7"`. It's one or the other.... but very soon just Python3 :smile:

bjlittle commented on .travis.yml:
```diff
@@ -16,7 +17,24 @@ install:
+        export ECCODES_DEPS="eccodes pip";
+    fi
+
+  - conda install --quiet --yes python=${python} iris numpy cartopy mock filelock pep8 $ECCODES_DEPS;
```
@lbdreyer Me being a shell snob - but could you make this `${ECCODES_DEPS}` please? :wink:

bjlittle commented on .travis.yml:
```diff
@@ -16,7 +17,24 @@ install:
+  # If running in Python 3, install eccodes-python and check the installation.
+  - if [[ "$python" == "3.6" ]]; then
+        pip install eccodes-python;
+        python -m eccodes selfcheck;
```
@lbdreyer Nice one :+1:

bjlittle commented on .travis.yml:
```diff
@@ -16,7 +17,24 @@ install:
+  - conda install --quiet --yes python=${python} iris numpy cartopy mock filelock pep8 $ECCODES_DEPS;
+    
+  # If running in Python 3, install eccodes-python and check the installation.
+  - if [[ "$python" == "3.6" ]]; then
```
@lbdreyer This could make this `"${python}" != "2.7"` instead, which is slightly more future proof and less tied to a specific minor version of Python 3

bjlittle commented on iris_grib/_save_rules.py:
```diff
@@ -976,7 +976,8 @@ def set_fixed_surfaces(cube, grib, full3d_cube=None):
             coeffs_array[n_coeffs + n_lev - 1] = sigma
         pv_values = [float(el) for el in coeffs_array]
-        gribapi.grib_set(grib, "NV", n_coeffs * 2)
+        # eccodes does not support writing numpy.it64, cast to python int
```
@lbdreyer Change `numpy.it64` -> `numpy.int64`

bjlittle commented on iris_grib/message.py:
```diff
@@ -33,9 +33,24 @@
+    is fixed.
+
+    """
+    return np.array(f)
```
@lbdreyer There a danger here that by only casting to `np.array` regardlessly we could lose a masked array that's passed in.

Should be be more selective?

lbdreyer replied:
You're right. We could either check its an array before casting, or perhaps use `np.asanyarray` which allows masked arrays to keep the mask:
```
>>> a  = ma.array([1,2,3], mask=[0,1,1])
>>> a
masked_array(data = [1 -- --],
             mask = [False  True  True],
       fill_value = 999999)

>>> f = np.asanyarray(a)
>>> f
masked_array(data = [1 -- --],
             mask = [False  True  True],
       fill_value = 999999)
```

bjlittle commented on iris_grib/message.py:
```diff
@@ -33,9 +33,24 @@
 _SUPPORTED_GRID_DEFINITIONS = (0, 1, 5, 10, 12, 20, 30, 40, 90)
 
 
+def grib_get_array_wrapper(f):
```
@lbdreyer I guess we should really make this a private function i.e. `_grib_get_array_wrapper`.

lbdreyer commented:
@bjlittle I've addressed your review actions. Do you mind taking another look?

bjlittle commented:
@lbdreyer Let's target this as a `0.14.1` release.

After merging this PR I created a `0.14.x` branch. Previously, I stuffed up and created a `0.14.x` branch for the `0.14.0` release but forget to update the docs (doh!), rather I done that on `master`.

Anyways, please use the new `0.14.x` branch for the `0.14.1` tag and release (which already includes this PR), and update the docs on `0.14.x` (although there may be issues building them on readthedocs due to eccodes not being available from PyPI - but don't let that block the release)

bjlittle commented:
Ping @rcomer :wink:

rcomer commented:
Great news!  Thanks both for your work on this :grin:
